### PR TITLE
Changes for end to end ssl and SSLPolicy in application gateways

### PR DESCRIPTION
**Pull request by vinatara**
The build will fail because the SDK package has not been published yet. Please look at changes and provide comments.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@vinatara__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (vinatara). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by markcowl** [comment]
@vinatara @vinayada Please coordinate, or neither of your PRs will be merged.

**Comment by markcowl** [comment]
@vinatara It looks like you need to publish the new version of the network package:  

EXEC : warning : Unable to find version '6.0.2-preview' of package 'Microsoft.Azure.Management.Network'. [D:\workspace\powershell\build.proj]

**Comment by markcowl** [comment]
Please make sure that this file is not a certifciate used for credentials anywhere:  ResourceManager/Network/Commands.Network.Test/ScenarioTests/Data/ApplicationGatewayAuthCert.cer

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Add, "AzureRmApplicationGatewayAuthenticationCertificate"), OutputType(typeof(PSApplicationGateway))]
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Add SupportsShouldProcess = true to the Cmdlet attribute

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+             Mandatory = true,
+             ValueFromPipeline = true,
+             HelpMessage = "The applicationGateway")]
+        public PSApplicationGateway ApplicationGateway { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+
+            var authCertificate = this.ApplicationGateway.AuthenticationCertificates.SingleOrDefault(resource => string.Equals(resource.Name, this.Name, System.StringComparison.CurrentCultureIgnoreCase));
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Guard this with a call to ConfirmAction(string target, string processMessage, Action executionAction)

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/NewAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, "AzureRmApplicationGatewayAuthenticationCertificate"), OutputType(typeof(PSApplicationGatewayAuthenticationCertificate))]
+    public class NewAzureApplicationGatewayAuthenticationCertificateCommand : AzureApplicationGatewayAuthenticationCertificateBase
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/NewAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Add SUpportsShouldProcess=true to the Cmdlet attribute

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/NewAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+            WriteObject(base.NewObject());
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/NewAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Protect this with a call to ConfirmAction(string, string, Action) or SHouldProcess.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/RemoveAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Remove, "AzureRmApplicationGatewayAuthenticationCertificate"), OutputType(typeof(PSApplicationGateway))]
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/RemoveAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Add SUpportsShouldProcess = true

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/RemoveAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+             Mandatory = true,
+             ValueFromPipeline = true,
+             HelpMessage = "The applicationGateway")]
+        public PSApplicationGateway ApplicationGateway { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+
+            var authCertificate = this.ApplicationGateway.AuthenticationCertificates.SingleOrDefault(resource => string.Equals(resource.Name, this.Name, System.StringComparison.CurrentCultureIgnoreCase));
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/RemoveAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Protect with a call to ConfirmAction(string, string,Action) or ShouldProcess.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/SetAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+using Microsoft.Azure.Commands.Network.Models;
+using System;
+using System.Linq;
+using System.Management.Automation;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Set, "AzureRmApplicationGatewayAuthenticationCertificate"), OutputType(typeof(PSApplicationGateway))]
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/SetAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Add SupportsShouldProcess = true

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/SetAzureApplicationGatewayAuthenticationCertificateCommand.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+             Mandatory = true,
+             ValueFromPipeline = true,
+             HelpMessage = "The applicationGateway")]
+        public PSApplicationGateway ApplicationGateway { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+
+            var oldAuthCertificate = this.ApplicationGateway.AuthenticationCertificates.SingleOrDefault(resource => string.Equals(resource.Name, this.Name, System.StringComparison.CurrentCultureIgnoreCase));
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/SetAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Protect execution with a call to ConfirmAction(string, string, Action) or a call to Shou;ldProcess

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslPolicy/NewAzureApplicationGatewaySslPolicyCommand.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, "AzureRmApplicationGatewaySslPolicy"), OutputType(typeof(PSApplicationGatewaySslPolicy))]
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslPolicy/NewAzureApplicationGatewaySslPolicyCommand.cs]: Same comments on SupportsShouldProcess and ConfirmAction/ShouldProcess

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslPolicyCommand.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+// ----------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Network.Models;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Set, "AzureRmApplicationGatewaySslPolicy"), OutputType(typeof(PSApplicationGateway))]
+    public class SetAzureApplicationGatewaySslPolicyCommand : AzureApplicationGatewaySslPolicyBase
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslPolicyCommand.cs]: Same comments on SupportsSHouldProcess and ConfirmAction/ShouldProcess

Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslPolicyCommand.cs]: Same comments on SupportsSHouldProcess and ConfirmAction/ShouldProcess

**Comment by markcowl** [comment]
@vinatara some comments.  See: https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d  For a detailed discussion of PowerShell Confirmation.  Note that the build will fail if you do not add help for the new cmdlets as well.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs (hunk 1 above)
Comment (vinatara) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs]: These are child resources of application gateway. I don't see SupportsShouldProcess  being added to any other child resources (e.g. AddAzureLoadBalancerBackendAddressPoolConfigCommand.cs for load balancer child resource). Also VS says it cannot resolve SupportsShouldProcess  in this file.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs (hunk 2 above)
Comment (vinatara) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/AuthenticationCertificate/AddAzureApplicationGatewayAuthenticationCertificateCommand.cs]: Same comment here. I don't see this being done for any other child resource. Is it needed?

File: src/Common/Commands.Common.Authentication/Commands.Common.Authentication.csproj [hunk 11]
```diff
@@ -15,6 +15,7 @@
     <RestorePackages>true</RestorePackages>
```
Comment (markcowl) [inline src/Common/Commands.Common.Authentication/Commands.Common.Authentication.csproj]: Revert any changes to this file

File: src/Common/Commands.Common.Storage/Commands.Common.Storage.csproj [hunk 12]
```diff
@@ -14,6 +14,7 @@
     <TargetFrameworkProfile />
```
Comment (markcowl) [inline src/Common/Commands.Common.Storage/Commands.Common.Storage.csproj]: Revert any changes to this file

File: src/Common/Commands.Common/Commands.Common.csproj [hunk 13]
```diff
@@ -15,6 +15,7 @@
     <RestorePackages>true</RestorePackages>
```
Comment (markcowl) [inline src/Common/Commands.Common/Commands.Common.csproj]: Revert any changes to this file

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Commands.ResourceManager.Common.csproj [hunk 14]
```diff
@@ -309,4 +309,11 @@
     </ProjectReference>
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/Commands.ResourceManager.Common.csproj]: Revert any changes to this file

File: src/ResourceManager/Common/Commands.ScenarioTests.ResourceManager.Common/Commands.ScenarioTests.ResourceManager.Common.csproj [hunk 15]
```diff
@@ -201,4 +201,11 @@
   </ItemGroup>
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ScenarioTests.ResourceManager.Common/Commands.ScenarioTests.ResourceManager.Common.csproj]: Revert

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 16]
```diff
@@ -434,4 +437,11 @@
   </ItemGroup>
   <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: I would remove this

File: src/ResourceManager/Tags/Commands.Tags/Commands.Tags.csproj [hunk 17]
```diff
@@ -171,4 +171,11 @@
     </ProjectReference>
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
```
Comment (markcowl) [inline src/ResourceManager/Tags/Commands.Tags/Commands.Tags.csproj]: Unneeded

File: src/ResourceManager/Resources/Commands.Resources/Commands.Resources.csproj [hunk 18]
```diff
@@ -246,4 +246,11 @@
   </ItemGroup>
   <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/Commands.Resources.csproj]: Unneeded, revert

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Commands.Resources.Rest.csproj [hunk 19]
```diff
@@ -286,4 +286,11 @@
   </ItemGroup>
   <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Commands.Resources.Rest.csproj]: revert

File: src/ResourceManager/Profile/Commands.Profile/Commands.Profile.csproj [hunk 20]
```diff
@@ -197,4 +197,11 @@
   </ItemGroup>
   <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
```
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/Commands.Profile.csproj]: revert

**Comment by markcowl** [comment]
on demand run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1063/

LGTM once the build passes

**Pull request merged**